crates/sui-indexer/src/handlers/objects_snapshot_processor.rs

```diff
@@ -221,26 +212,43 @@ impl ObjectsSnapshotProcessor {
                         .await?
                         .unwrap_or_default();
 
+                    // Fetch new changes if buffer is empty
+                    if buffered_changes.is_empty() {
+                        if let Some(new_changes) = stream.next().await {
+                            buffered_changes.extend(new_changes);
+                            // Can't really guarantee order, so we sort checkpoints ascending here
+                            buffered_changes.sort_by_key(|change| change.checkpoint_sequence_number);
+                        }
+                    }
+
                     // We update the snapshot table when it falls behind the rest of the indexer by more than the min_lag.
-                    while latest_indexer_cp >= start_cp + config.snapshot_min_lag as u64 {
-                        // Stream the next object changes to be committed to the store.
-                        if let Some(object_changes_batch) = stream.next().await {
-                            let first_checkpoint_seq = object_changes_batch.first().as_ref().unwrap().checkpoint_sequence_number;
-                            let last_checkpoint_seq = object_changes_batch.last().as_ref().unwrap().checkpoint_sequence_number;
-                            info!("Objects snapshot processor is updating objects snapshot table from {} to {}", first_checkpoint_seq, last_checkpoint_seq);
-
-                            let changes_to_commit = object_changes_batch.into_iter().map(|obj| obj.object_changes).collect();
-                            store.backfill_objects_snapshot(changes_to_commit)
-                                .await
-                                .unwrap_or_else(|_| panic!("Failed to backfill objects snapshot from {} to {}", first_checkpoint_seq, last_checkpoint_seq));
-                            start_cp = last_checkpoint_seq + 1;
-
-                            // Tells the package buffer that this checkpoint has been processed and the corresponding package data can be deleted.
-                            commit_notifier.send(Some(last_checkpoint_seq)).expect("Commit watcher should not be closed");
-                            metrics
-                                .latest_object_snapshot_sequence_number
-                                .set(last_checkpoint_seq as i64);
+                    while latest_indexer_cp >= start_cp + config.lag as u64 && !buffered_changes.is_empty() {
+                        let max_allowed_cp = latest_indexer_cp - config.lag as u64;
+
+                        let split_index = buffered_changes.partition_point(|change| change.checkpoint_sequence_number <= max_allowed_cp);
+
+                        if split_index == 0 {
```

Comment:
after 2 comments above, this check seems un-necessary too, b/c we know that 1. buffer is not empty 2. buffer's first is start_cp + 1 3. start_cp + 1 <= max_allowed_cp
thus there is at least 1 item to be persisted.